buildtest/buildsystem/base.py

```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+    # https://software.intel.com/content/www/us/en/develop/documentation/fortran-compiler-developer-guide-and-reference/top/compiler-setup/using-the-command-line/understanding-file-extensions.html
+    # Fortran Extensions: http://fortranwiki.org/fortran/show/File+extensions
+    lang_ext_table = {
+        ".c": "C",
+        ".cc": "C++",
+        ".cxx": "C++",
+        ".cpp": "C++",
+        ".c++": "C++",
+        ".f90": "Fortran",
+        ".F90": "Fortran",
+        ".f95": "Fortran",
+        ".f03": "Fortran",
+        ".f": "Fortran",
+        ".F": "Fortran",
+        ".FOR": "Fortran",
+        ".for": "Fortran",
+        ".FTN": "Fortran",
+        ".ftn": "Fortran",
+    }
+
+    cc = None
+    cxx = None
+    fc = None
+    ldflags = None
+    cflags = None
+    cxxflags = None
+    fflags = None
+    cppflags = None
+
+    def _build_testcontent(self):
+        """This method will build the test content from a Buildspec that uses compiler schema. We need a 'compiler'
+           and 'source' key which specifies the source files to compile. We resolve the source file path which can
+           be an absolute value or relative path with respect to Buildspec. The file extension of sourcefile is used
+           to detect the Programming Language which is used to lookup the compiler wrapper based on Language + Compiler.
+           During compiler detection, we set class variables ``self.cc``, ``self.cxx``. ``self.fc``, ``self.cflags``,
+           ``self.cxxflags``, ``self.fflags``, ``self.cppflags``. ``self.ldflags``. Finally we generate the compile
+           command and add each instruction to ``lines`` which contains content of test. Upon completion, we return
+           a list that contains content of the test.
+        """
+
+        self.compiler_recipe = self.recipe.get("compiler")
+        sourcefile = self.compiler_recipe["source"]
+        # The 'source' key can accept an absolute path or relative path with respect to Buildspec file.
+        # resolve_path returns full path to file or None if not found. We first check absolute path if not found we
+        # check for relative path.
+        self.source = resolve_path(sourcefile) or resolve_path(
+            os.path.join(
+                os.path.dirname(self.buildspec), self.compiler_recipe["source"]
+            )
+        )
+
+        # raise error if we can't find source file to compile
+        if not self.source:
+            sys.exit(f"Failed to resolve path specified by key 'source': {sourcefile}")
+
+        # name of generated executable. Add the .exe extension at end of source file name.
+        self.executable = "%s.exe" % os.path.basename(self.source)
+
+        self.compiler_name = None
+        self.lang = self.detect_lang(self.source)
+
+        # Get GNU Compiler if 'gnu' defined in compiler properties.
+        if self.compiler_recipe.get("gnu"):
+            self.compiler_name = "gnu"
+            gnu = GNUCompiler()
```

Comment:
Okay here is where I still think it needs work - right now you are passing _all_ kinds of compiler specs into the Compiler class, and doing this complicated if/then/else thing. The executors should be modular (and use different names as necessary) so we start, off the bat, with the recipe mapped to the right compiler. The Compiler class does not exist to be a jump off point to determine which other class to instantiate, but instead an Abstract class to implement common functions or variables. Take a look at snakemake.executors if you need a good example of modularity with parent/child classes.